src/main/java/uk/gov/ons/ctp/response/casesvc/service/CaseService.java

```diff
@@ -65,91 +195,411 @@
-   * a side effect create a new case - if so the details must be provided in the newCase argument,
-   * otherwise it may remain null. If the newCase is passed it must also contain the contact details
-   * for the new case.
+   * Overwrites existing iac for given case if iac is not active and adds new iac to caseiacaudit
+   * table
+   *
+   * @param targetCase Case to update
+   */
+  private void generateAndStoreNewIAC(final Case targetCase) {
+    String iac = targetCase.getIac();
+    if (iac == null || !internetAccessCodeSvcClient.isIacActive(iac)) {
+      log.with("case_id", targetCase.getId()).debug("Replacing existing case IAC");
+      String newIac = internetAccessCodeSvcClient.generateIACs(1).get(0);
+      targetCase.setIac(newIac);
+      caseRepo.saveAndFlush(targetCase);
+      saveCaseIacAudit(targetCase);
+    } else {
+      log.with("case_id", targetCase.getId()).debug("Existing IAC is still active");
+    }
+  }
+
+  private void transitionCaseGroupStatus(final Case targetCase, final CaseEvent caseEvent) {
+    CaseGroup caseGroup = caseGroupRepo.findOne(targetCase.getCaseGroupFK());
+    try {
+      caseGroupService.transitionCaseGroupStatus(
+          caseGroup, caseEvent.getCategory(), targetCase.getPartyId());
+    } catch (CTPException e) {
+      // The transition manager throws an exception if the event doesn't cause a transition, however
+      // there are lots of
+      // events which do not cause CaseGroupStatus transitions, (this is valid behaviour).
+      log.debug(e.getMessage());
+    }
+  }
+
+  private void processActionPlanChange(final Case targetCase, final boolean enrolments)
+      throws CTPException {
+
+    List<CaseGroup> caseGroups =
+        caseGroupService.findCaseGroupsForExecutedCollectionExercises(targetCase);
+
+    for (CaseGroup caseGroup : caseGroups) {
+
+      // fetch all B and BI cases associated to the case group being processed
+      List<Case> cases =
+          caseRepo.findByCaseGroupFKOrderByCreatedDateTimeDesc(caseGroup.getCaseGroupPK());
+
+      List<ActionPlanDTO> actionPlans =
+          actionSvcClient.getActionPlans(caseGroup.getCollectionExerciseId(), enrolments);
+
+      if (actionPlans == null || actionPlans.size() != 1) {
+        log.with("collection_exercise_id", caseGroup.getCollectionExerciseId())
+            .with("enrolments", enrolments)
+            .error("One action plan expected");
+        throw new IllegalStateException(
+            "Expected one action plan for collection exercise with enrolmentStatus");
+      }
+
+      for (Case caze : cases) {
+        if (caze.getSampleUnitType() == SampleUnitType.B) {
+          caze.setActionPlanId(actionPlans.get(0).getId());
+          caseRepo.saveAndFlush(caze);
+          notificationPublisher.sendNotification(
+              prepareCaseNotification(caze, CaseDTO.CaseEvent.ACTIONPLAN_CHANGED));
+        }
+      }
+    }
+  }
+
+  @Transactional(
+      propagation = Propagation.REQUIRED,
```

Comment:
We don't need all the extra properties specified. `@Transactional` should be plenty